src/sshfs_mount/sshfs_mount_handler.cpp

```diff
@@ -60,229 +48,162 @@ void start_and_block_until(mp::Process* process, Signal signal, std::function<bo
-    {
-        throw mp::SSHFSMissingError();
-    }
+        throw SSHFSMissingError();
     else if (process_state.exit_code || process_state.error)
-    {
         throw std::runtime_error(
-            fmt::format("{}: {}", process_state.failure_message(), sshfs_server_process->read_all_standard_error()));
-    }
-
-    sshfs_server_configs[vm->vm_name].erase(target_path);
-    mount_processes[vm->vm_name][target_path] = std::move(sshfs_server_process);
+            fmt::format("{}: {}", process_state.failure_message(), process->read_all_standard_error()));
 }
 
-void mp::SSHFSMountHandler::stop_mount(const std::string& instance, const std::string& path)
+void SSHFSMountHandler::stop()
 {
-    auto sshfs_mount_it = mount_processes.find(instance);
-    if (sshfs_mount_it == mount_processes.end())
-    {
-        mpl::log(mpl::Level::info, category,
-                 fmt::format("No running mount process for \"{}\" serving '{}'", instance, path));
+    if (!(process && process->running()))
         return;
-    }
-
-    auto& sshfs_mount_map = sshfs_mount_it->second;
-    auto map_entry = sshfs_mount_map.find(path);
-    if (map_entry != sshfs_mount_map.end())
-    {
-        auto& sshfs_mount = map_entry->second;
-        mpl::log(mpl::Level::info, category, fmt::format("Stopping mount '{}' in instance \"{}\"", path, instance));
-        sshfs_mount->terminate();
 
-        if (!sshfs_mount->wait_for_finished(1000))
-        {
-            mpl::log(mpl::Level::info, category,
-                     fmt::format("Failed to terminate mount '{}' in instance \"{}\", killing", path, instance));
-            sshfs_mount->kill();
-        }
-    }
+    mpl::log(mpl::Level::info, category, fmt::format("Stopping mount \"{}\" in instance '{}'", target, vm->vm_name));
+    if (process->terminate(); !process->wait_for_finished(5000))
+        throw std::runtime_error(
+            fmt::format("Failed to terminate SSHFS mount process: {}", process->read_all_standard_error()));
 }
-
-void mp::SSHFSMountHandler::stop_all_mounts_for_instance(const std::string& instance)
+SSHFSMountHandler::~SSHFSMountHandler()
 {
-    auto mounts_it = mount_processes.find(instance);
-    if (mounts_it == mount_processes.end() || mounts_it->second.empty())
+    try
     {
-        mpl::log(mpl::Level::info, category, fmt::format("No mounts to stop for instance \"{}\"", instance));
+        stop();
     }
-    else
+    catch (const std::exception& e)
     {
-        for (auto it = mounts_it->second.cbegin(); it != mounts_it->second.cend();)
-        {
-            // Clever postfix increment with member access needed to prevent iterator invalidation since iterable is
-            // modified in sshfs connection signal handler
-            stop_mount(instance, it++->first);
-        }
+        mpl::log(
+            mpl::Level::warning, category,
+            fmt::format("Failed to gracefully stop mount \"{}\" in instance '{}': {}", target, vm->vm_name, e.what()));
```

Comment:
You need to kill the process at this point, right? Otherwise, slots might be called after it was deleted.

Reply:
The destructor of `QProcess` kills the process, so it should be killed immediately after this point.

Reply:
Ahh right! Good. I wonder why there was that comment about the delete_later in the code...

Reply:
From my understanding, it was based on a false assumption. It stated that the slots connected to the process signals could be called on the event loop, so in order to ensure that by the time they are called, the process is not destroyed, it used a `qt_delete_later_unique_ptr`, so that the process is also destroyed on the event loop, after all the slots have been called. But this should never be the case. The concern about the slots being called on the event loop is valid only for `Qt::QueuedConnection`. But when connecting a lambda as a slot to a signal, the connection is always `Qt::DirectConnection`, which means the slots will be called immediately, and the `emit` call will only return after all the slots have been called. That's why now the mount handler uses a simple `unique_ptr` to hold the process. By the time it is destroyed, it can't have any slots waiting to be called.